Warn if an item doesn't have a link from every child document

michaelnt opened the pull request:
Fixes issue #141.

Let me know what you'd like changed to get this merged.

jacebrowning commented on doorstop/cli/main.py:
```diff
@@ -52,6 +52,8 @@ def main(args=None):  # pylint: disable=R0915
     parser.add_argument('-C', '--no-child-check', action='store_true',
                         help="do not validate child (reverse) links")
+    parser.add_argument('-Z', '--strict-child-check', action='store_true',
+                        help="Strictly validate (reverse) links from every child document")
```
To me, a help message like "require child (reverse) links from every document" is more clear.

jacebrowning commented on doorstop/settings.py:
```diff
@@ -30,6 +30,7 @@
 CHECK_LEVELS = True  # validate document levels during validation
 CHECK_REF = True  # validate external file references
 CHECK_CHILD_LINKS = True  # validate reverse links
+CHECK_CHILD_LINKS_STRICT = False  # Strictly validate reverse links from all the child documents
```
To me, a comment like "require reverse links from every child document" is more clear.

jacebrowning commented on doorstop/core/item.py:
```diff
@@ -652,7 +653,7 @@ def _get_issues_tree(self, tree):
             self._data['links'] = identifiers
 
-    def _get_issues_both(self, document, tree, skip):
+    def _get_issues_both(self, document, tree, skip, children):
```
Is there any way this could be accomplished without passing `children` as a argument? This method already has a reference to the tree, so the child documents could be determined inside the `if settings.CHECK_CHILD_LINKS_STRICT:` block. I'm concerned about this breaking other functionality and showing down validation in cases were `CHECK_CHILD_LINKS_STRICT` is not needed.

michaelnt commented:
I changed the help string and settings comment. 

I removed the children argument but unfortunately document.tree points to the root of the tree so I added document.children that calls self.tree.get_prefix_of_children() to return the tree below a document. Perhaps that code should be in just tree or document.

michaelnt commented on doorstop/core/item.py:
```diff
@@ -664,7 +665,8 @@ def _get_issues_both(self, document, tree, skip):
             items, documents = self._find_child_objects(document=document,
                                                         tree=tree,
-                                                        find_all=False)
+                                                        find_all=True)
```
Should find_all only be run if CHECK_CHILD_LINKS_STRICT is true to speed things up?

jacebrowning replied:
Yeah, I think so.

michaelnt replied:
I've made that change

jacebrowning commented on doorstop/core/tree.py:
```diff
@@ -437,7 +438,12 @@ def get_issues(self, skip=None, document_hook=None, item_hook=None):
             yield DoorstopWarning("no documents")
         # Check each document
-        for document in documents:
+        trees = [self]
```
What is this logic doing? I'm a bit confused to see a variable named `trees` -- is there more than one?

michaelnt replied:
Sorry that isn't needed anymore and I've removed this change

michaelnt commented:
Rather than updating document.children when ever it is read it is now set when the tree is modified. I used the method _place, is that the right place to put the code?

michaelnt commented on doorstop/core/tree.py:
```diff
@@ -477,6 +479,16 @@ def by_uid(row):
+                document.children = children
+                return
+        children = [c.document.prefix for c in self.children]
+        document.children = children
```
OK, thats changed

jacebrowning commented:
I'm going to make a beta release for this. Feel free to make a new PR with any tweaks.
